README.md
<p align="center">
  <br />
  <img width="150" height="150" src="./dictionary.png" alt="Logo">
  <h1 align="center"><b>wordlist-js</b></h1>
  <div align="center">
    <a href="https://www.typescriptlang.org/">
      <img src="https://img.shields.io/badge/Language-Typescript-%233178C6.svg?style=flat&logo=typescript" alt="Language: TypeScript">
    </a>
    <a href="https://www.npmjs.com/package/wordlist-js">
      <img src="https://img.shields.io/npm/v/wordlist-js" alt="NPM version">
    </a>
    <a href="https://www.npmjs.com/package/wordlist-js">
      <img src="https://img.shields.io/bundlephobia/min/wordlist-js" alt="NPM bundle size">
    </a>
    <a href="https://www.npmjs.com/package/wordlist-js">
      <img src="https://img.shields.io/npm/dw/wordlist-js" alt="NPM downloads">
    </a>
    <a href="https://github.com/jordanshatford/wordlist-js/blob/main/LICENSE">
      <img src="https://img.shields.io/badge/License-MIT-black.svg?style=flat&logo=license" alt="License: MIT">
    </a>
  </div>
  <p align="center">
    Lists of filtered words from various English dialects.
    <br />
    <a href="https://www.npmjs.com/package/wordlist-js"><strong>www.npmjs.com/package/wordlist-js »</strong></a>
    <br />
    <br />
  </p>
</p>

wordlist-js provides lists of English words based on how common they are used. It also provides words for many English dialects.

## Available Words Lists

See the project [config](./wordlist.config.json) for information on dialects and frequencies generated. Alternatively see the [dist](./dist/) folder for type declaration files.

## SCOWL

This project and words lists are based on [SCOWL](http://wordlist.aspell.net/).

When manually combining the words lists the "english" spelling category should be used as well as one of "american", "british", "britishZ" (british with ize spelling), "canadian" or "australian".

Size 35 is the recommended small size, 50 the medium and 70 the large. Sizes 70 and below contain words found in most dictionaries while the 80 size contains all the strange and unusual words people like to use in word games such as Scrabble (TM). While a lot of the words in the 80 size are not used very often, they are all generally considered valid words in the English language. The 95 contains just about every English word in existence and then some. Many of the words at the 95 level will probably not be considered valid English words by most people.

For spell checking it is recommend to use size 60. The 70 size is reasonable for those wanting a larger list and don't mind a few errors. The 80 or larger sizes are not reasonable for spell checking.


CHANGELOG.md
# wordlist-js

## 2.0.0

### Major Changes

- breaking: update exports to specify filtered out lists ([`8142c376b79c153169692664df64a99532ee855e`](https://github.com/jordanshatford/wordlist-js/commit/8142c376b79c153169692664df64a99532ee855e))

### Minor Changes

- feat: better generation of filtered and filtered out words lists ([`07760920648bb0120ba3dc8d000f0fd3bb5b2fac`](https://github.com/jordanshatford/wordlist-js/commit/07760920648bb0120ba3dc8d000f0fd3bb5b2fac))

- feat: export lists of words unfiltered ([`2e915f21cc6b98c6d4f63237d123c230fb05f9c6`](https://github.com/jordanshatford/wordlist-js/commit/2e915f21cc6b98c6d4f63237d123c230fb05f9c6))

### Patch Changes

- fix: better handling of invalid dialects and frequencies when generating ([`acf0f4f24c30cbb04aa4ca6117e6047ab9e2f267`](https://github.com/jordanshatford/wordlist-js/commit/acf0f4f24c30cbb04aa4ca6117e6047ab9e2f267))

## 1.8.0

### Minor Changes

- feat: add helper function for generating array of elements export ([`56dd8fd`](https://github.com/jordanshatford/wordlist-js/commit/56dd8fd4be0bee567baf006d167e3f168cf40501))

- feat: generate list of all bad words for each dialect ([`293544c`](https://github.com/jordanshatford/wordlist-js/commit/293544c00380146be644003fe5c692a966064ee4))

- feat: use typescript compiler api to generate dist code ([`9d96a90`](https://github.com/jordanshatford/wordlist-js/commit/9d96a900c1a6f8721bd311abdff20cda7562a59f))

- feat: simplify words list processing ([`27c7923`](https://github.com/jordanshatford/wordlist-js/commit/27c7923f6501fab6a1623da1f7c4881075be7544))

## 1.7.1

### Patch Changes

- fix: properly export from each dialect causing breaking change ([`8152abf`](https://github.com/jordanshatford/wordlist-js/commit/8152abfeb896ff815c7daa7f0b943b3a4649f321))

## 1.7.0

### Minor Changes

- feat: store words lists as json and get rid of scowl (see [#96](https://github.com/jordanshatford/wordlist-js/issues/96)) ([`0af1303`](https://github.com/jordanshatford/wordlist-js/commit/0af130327d18acdd998adb081839c765d61a1ede))

- feat: simply checking if word is in bad word list ([`6854893`](https://github.com/jordanshatford/wordlist-js/commit/6854893eebebe4e54d4864f0613e14858ba645d9))

- feat: only filter out exact words listed instead of regexs ([`594abdc`](https://github.com/jordanshatford/wordlist-js/commit/594abdc619a907a8b524544bcdd04aa8e770e3ab))

- feat: export bad words separately (see [#92](https://github.com/jordanshatford/wordlist-js/issues/92)) ([`277f02c`](https://github.com/jordanshatford/wordlist-js/commit/277f02ccbbeb500cbbdb80343b997ddcccd46505))

- feat: use wordlist.config.json when generating build ([`822e247`](https://github.com/jordanshatford/wordlist-js/commit/822e247def5516fd76d3db086ae41f919e4421db))

- feat: replace bad-words with our own list (as it is unmaintained) ([`f050336`](https://github.com/jordanshatford/wordlist-js/commit/f050336d5b6d39b94f4d4194276c413149467314))

- feat: better bad word matching to cleanup many missed words ([`932aa75`](https://github.com/jordanshatford/wordlist-js/commit/932aa7525d4ed7876e71503d8eac35672a068eb2))

### Patch Changes

- fix: remove duplicates from bad words list ([`8e0fdae`](https://github.com/jordanshatford/wordlist-js/commit/8e0fdaec7ea1b24c000979f8bbd99e840bfbce43))

## 1.6.0

### Minor Changes

- feat: add release ci workflow using changesets ([`2c4ed28`](https://github.com/jordanshatford/wordlist-js/commit/2c4ed28572b073987e0f83b7b937d55c7fcaa7e4))

- feat: use pnpm instead of npm (see [#99](https://github.com/jordanshatford/wordlist-js/issues/99)) ([`11c5b7f`](https://github.com/jordanshatford/wordlist-js/commit/11c5b7f1707982d9551403ec286978e64e479430))

### Patch Changes

- chore: replace semantic-release with changesets ([`132bc64`](https://github.com/jordanshatford/wordlist-js/commit/132bc64d4a1b7d320006089c93202744636c767b))


scripts/generate.ts
import fs from 'fs';
import path from 'path';
import ts from 'typescript';
import config from '../wordlist.config.json';

const ALLOWED_DIALECTS = ['english', 'american', 'australian', 'british', 'canadian', 'britishZ'];
const ALLOWED_FREQUENCIES = [10, 20, 35, 40, 50, 55, 60, 70, 80, 95];

const FILTERED_SUFFIX = '';
const FILTERED_OUT_SUFFIX = 'FilteredOut';
const UNFILTERED_SUFFIX = 'Unfiltered';

const projectDir = path.join(__dirname, '..');

/**
 * Process a list of words. Filter out specific words.
 * @param words - the list of words.
 * @param toFilter - the words to filter out.
 * @returns = object containing list of filtered and filtered out words.
 */
export function processWordsList(
  words: string[],
  toFilter: string[]
): { filtered: string[]; filteredOut: string[] } {
  // Sort list of words.
  const sortedWords = words.sort();
  // Track all words filtered out.
  const filteredOut: string[] = [];
  const filteredWords = sortedWords.filter((word) => {
    const shouldFilter = toFilter.some((w) => w.toLowerCase() === word.toLowerCase());
    if (shouldFilter) {
      filteredOut.push(word);
    }
    return !shouldFilter;
  });
  return { filtered: filteredWords, filteredOut };
}

export function generate() {
  const outputDirPath = path.join(projectDir, config.output);
  const sourceDirPath = path.join(projectDir, config.source);

  fs.mkdir(outputDirPath, { recursive: true }, (err) => {
    if (err) throw err;
  });

  const filteredWords = fs.readFileSync(path.join(sourceDirPath, '_filtered.json'), 'utf-8');
  const splitFilteredWords: string[] = JSON.parse(filteredWords);

  // Iterate over each dialect to create a specific file for that dialect.
  for (const dialect of config.dialects) {
    // If the dialect isnt supported ignore and continue.
    if (!ALLOWED_DIALECTS.includes(dialect)) {
      continue;
    }
    const content = fs.readFileSync(path.join(sourceDirPath, `${dialect}.json`), 'utf-8');
    const json: Record<string, string[]> = JSON.parse(content);
    const info: Array<{ name: string; filtered: string[]; filteredOut: string[] }> = [];
    // Iterate over each frequency for the dialect and add information to be generated in file.
    for (const frequency of config.frequencies) {
      // If the frequency isnt supported ignore and continue.
      if (!ALLOWED_FREQUENCIES.includes(frequency)) {
        continue;
      }
      const words = json[frequency];
      const { filtered, filteredOut } = processWordsList(words, splitFilteredWords);
      info.push({
        name: `${dialect}${frequency}`,
        filtered,
        filteredOut,
      });
    }
    // Generate `dialect`.ts file for the current dialect.
    createDialectFile(dialect, info, outputDirPath);
  }
  // Generate index.ts file linking all dialect files.
  createIndexFile(config.dialects, outputDirPath);
}

/**
 * Create file for a specific dialect.
 * @param dialect - the dialect.
 * @param frequencies - the frequencies for that dialect.
 * @param outDir - the directory to output the file to.
 */
function createDialectFile(
  dialect: string,
  frequencies: { name: string; filtered: string[]; filteredOut: string[] }[],
  outDir: string
) {
  const statementArray: ts.Statement[] = [];
  // Add array for each frequency.
  for (const f of frequencies) {
    const frequencyArray = createExportedConstStringArray(
      `${f.name}${FILTERED_SUFFIX}`,
      f.filtered.map((w) => ts.factory.createStringLiteral(w))
    );
    statementArray.push(frequencyArray);
    const unfilteredItems = [`${f.name}${FILTERED_SUFFIX}`];
    if (f.filteredOut.length > 0) {
      const frequencyArrayFilteredOut = createExportedConstStringArray(
        `${f.name}${FILTERED_OUT_SUFFIX}`,
        f.filteredOut.map((w) => ts.factory.createStringLiteral(w))
      );
      statementArray.push(frequencyArrayFilteredOut);
      unfilteredItems.push(`${f.name}${FILTERED_OUT_SUFFIX}`);
    }
    const dialectUnfiltered = createExportedConstStringArray(
      `${f.name}${UNFILTERED_SUFFIX}`,
      unfilteredItems.map((i) => ts.factory.createSpreadElement(ts.factory.createIdentifier(i)))
    );
    statementArray.push(dialectUnfiltered);
  }
  // Add array for all (filtered words) of the dialect.
  const all = createExportedConstStringArray(
    `${dialect}All${FILTERED_SUFFIX}`,
    frequencies
      .filter((f) => f.filtered.length > 0)
      .map((f) => ts.factory.createSpreadElement(ts.factory.createIdentifier(f.name)))
  );
  statementArray.push(all);
  // List of all filtered out words
  const allFilteredOut = createExportedConstStringArray(
    `${dialect}All${FILTERED_OUT_SUFFIX}`,
    frequencies
      .filter((f) => f.filteredOut.length > 0)
      .map((f) =>
        ts.factory.createSpreadElement(
          ts.factory.createIdentifier(`${f.name}${FILTERED_OUT_SUFFIX}`)
        )
      )
  );
  statementArray.push(allFilteredOut);
  // List of all words Unfiltered
  const allNotFiltered = createExportedConstStringArray(`${dialect}All${UNFILTERED_SUFFIX}`, [
    ts.factory.createSpreadElement(ts.factory.createIdentifier(`${dialect}All${FILTERED_SUFFIX}`)),
    ts.factory.createSpreadElement(
      ts.factory.createIdentifier(`${dialect}All${FILTERED_OUT_SUFFIX}`)
    ),
  ]);
  statementArray.push(allNotFiltered);
  createTypeScriptFile(outDir, `${dialect}.ts`, statementArray);
}

/**
 * Create an exported const array with provided name and elements.
 * @param name - name of the export.
 * @param elements - elements in the export.
 * @returns ts.VariableStatement - the variable statement for the array.
 */
function createExportedConstStringArray(
  name: string,
  elements: ts.Expression[]
): ts.VariableStatement {
  return ts.factory.createVariableStatement(
    ts.factory.createModifiersFromModifierFlags(ts.ModifierFlags.Export),
    ts.factory.createVariableDeclarationList(
      [
        ts.factory.createVariableDeclaration(
          name,
          undefined,
          ts.factory.createArrayTypeNode(
            ts.factory.createKeywordTypeNode(ts.SyntaxKind.StringKeyword)
          ),
          ts.factory.createArrayLiteralExpression(elements)
        ),
      ],
      ts.NodeFlags.Const
    )
  );
}

/**
 * Create the index file used in the dist.
 * @param files - list of files to include.
 * @param outDir - the directory to output the file to.
 */
function createIndexFile(files: string[], outDir: string) {
  const statementArray: ts.Statement[] = [];
  for (const f of files) {
    const moduleSpec = ts.factory.createStringLiteral(`./${f}`);
    // Add import declaration.
    const importDeclaration = ts.factory.createImportDeclaration(
      undefined,
      ts.factory.createImportClause(
        false,
        undefined,
        ts.factory.createNamespaceImport(ts.factory.createIdentifier(f))
      ),
      moduleSpec
    );
    statementArray.push(importDeclaration);
    // Export declaration.
    const exportDeclaration = ts.factory.createExportDeclaration(
      undefined,
      false,
      undefined,
      moduleSpec
    );
    statementArray.push(exportDeclaration);
  }
  // Creating a default export statement
  const exportDefault = ts.factory.createExportDefault(
    ts.factory.createObjectLiteralExpression([
      ...files.map((f) => {
        return ts.factory.createSpreadAssignment(ts.factory.createIdentifier(f));
      }),
    ])
  );
  statementArray.push(exportDefault);
  createTypeScriptFile(outDir, 'index.ts', statementArray);
}

/**
 * Create a TypeScript file on the filesystem.
 * @param dir - Directory for the file.
 * @param name - Name of the file.
 * @param statements - Statements to include in the file.
 */
function createTypeScriptFile(dir: string, name: string, statements: ts.Statement[]) {
  const file = ts.createSourceFile(name, '', ts.ScriptTarget.ESNext);
  const fileWithStatements = ts.factory.updateSourceFile(file, statements);
  const printer = ts.createPrinter({ newLine: ts.NewLineKind.LineFeed });
  const code = printer.printFile(fileWithStatements);
  fs.writeFileSync(path.join(dir, file.fileName), code);
}

generate();
